### PR TITLE
ESP32: Map ESP32 error codes with CHIP error codes

### LINKED ISSUE #4256: Map ESP32 error codes with CHIP error codes
#### Problem
ESP32 APIs return the error codes specific to ESP32. When these are called in the CHIP code, they should be mapped to corresponding CHIP error codes to be propagated forward so that the behavior is correct

 #### Proposed Solution
Add a function that converts ESP32 error codes to CHIP error codes and vice versa (if required)

### PR DESCRIPTION
**Pull request by sweetymhaiske**
#### Problem

ESP32 APIs return the error codes specific to ESP32. When these API are called in the CHIP code, they should be mapped to corresponding CHIP error codes to be propagated forward so that the behavior is correct

 #### Summary of Changes

Added the function which maps ESP32 APIs return codes to CHIP error codes.

 Fixes #4256

### CODE REVIEW

File: src/platform/ESP32/BLEManagerImpl.h [hunk 1]
```diff
@@ -217,6 +217,7 @@ class BLEManagerImpl final : public BLEManager,
     void HandleDisconnect(esp_ble_gatts_cb_param_t * param);
     CHIPoBLEConState * GetConnectionState(uint16_t conId, bool allocate = false);
     bool ReleaseConnectionState(uint16_t conId);
+    CHIP_ERROR MapBTError(int btErr);
```
Comment (dhrishi) [inline src/platform/ESP32/BLEManagerImpl.h:220]: Why is this required? Can't there be a single function `MapBLEError` which is used for both Bluedroid and NimBLE?

File: src/platform/ESP32/CHIPDevicePlatformConfig.h [hunk 2]
```diff
@@ -29,6 +29,8 @@
 
 #define CHIP_DEVICE_CONFIG_LWIP_WIFI_STATION_IF_NAME "st"
 
+#define CHIP_DEVICE_CONFIG_ESP32_BLE_ERROR_MIN 1100000
+#define CHIP_DEVICE_CONFIG_ESP32_BT_ERROR_MIN 1200000
```
Comment (dhrishi) [inline src/platform/ESP32/CHIPDevicePlatformConfig.h:33]: Here as well, either use a common macro (as only either of Bluedroid or NimBLE will be enabled at one time) or use NIMBLE / BLUEDROID in the macro name

**Review by dhrishi — CHANGES_REQUESTED** [review]

File: src/platform/ESP32/BLEManagerImpl.h (hunk 1 above)
Comment (sweetymhaiske) [reply src/platform/ESP32/BLEManagerImpl.h]: Done. Added the single function `MapBLEError` for both.

File: src/platform/ESP32/CHIPDevicePlatformConfig.h (hunk 2 above)
Comment (sweetymhaiske) [reply src/platform/ESP32/CHIPDevicePlatformConfig.h]: Done. Kept the common macro `CHIP_DEVICE_CONFIG_ESP32_BLE_ERROR_MIN` .

**Review by woody-apple — APPROVED** [review]

File: src/platform/ESP32/bluedroid/BLEManagerImpl.cpp [hunk 3]
```diff
@@ -347,12 +347,14 @@ bool BLEManagerImpl::UnsubscribeCharacteristic(BLE_CONNECTION_OBJECT conId, cons
...
 bool BLEManagerImpl::CloseConnection(BLE_CONNECTION_OBJECT conId)
 {
+    int ret;
     CHIP_ERROR err;
 
     ChipLogProgress(DeviceLayer, "Closing BLE GATT connection (con %u)", conId);
 
     // Signal the ESP BLE layer to close the conntion.
-    err = esp_ble_gatts_close(mAppIf, conId);
+    ret = esp_ble_gatts_close(mAppIf, conId);
```
Comment (yufengwangca) [inline src/platform/ESP32/bluedroid/BLEManagerImpl.cpp:356]: ret is not used for checking the result, this can be saved.


    CHIP_ERROR err = CHIP_NO_ERROR;

    ChipLogProgress(DeviceLayer, "Closing BLE GATT connection (con %u)", conId);

    // Signal the ESP BLE layer to close the conntion.

    err = MapBLEError(esp_ble_gatts_close(mAppIf, conId));
    if (err != CHIP_NO_ERROR)
    {
        ChipLogError(DeviceLayer, "esp_ble_gatts_close() failed: %s", ErrorStr(err));
    }

File: src/platform/ESP32/bluedroid/BLEManagerImpl.cpp [hunk 4]
```diff
@@ -378,6 +380,7 @@ uint16_t BLEManagerImpl::GetMTU(BLE_CONNECTION_OBJECT conId) const
 bool BLEManagerImpl::SendIndication(BLE_CONNECTION_OBJECT conId, const ChipBleUUID * svcId, const ChipBleUUID * charId,
                                     PacketBufferHandle data)
 {
+    int ret;
```
Comment (yufengwangca) [inline src/platform/ESP32/bluedroid/BLEManagerImpl.cpp:383]: Same, this variable can be saved

File: src/platform/ESP32/nimble/BLEManagerImpl.cpp [hunk 5]
```diff
@@ -426,6 +430,35 @@ bool BLEManagerImpl::SendReadResponse(BLE_CONNECTION_OBJECT conId, BLE_READ_REQU
...
+    case BLE_HS_EBADDATA:
+        return CHIP_ERROR_DATA_NOT_ALIGNED;
+    case BLE_HS_ETIMEOUT:
+        return CHIP_ERROR_TIMEOUT;
+    case BLE_HS_ENOADDR:
+        return CHIP_ERROR_INVALID_ADDRESS;
+    case ESP_ERR_INVALID_ARG:
+        return CHIP_ERROR_INVALID_ARGUMENT;
+    default:
+        return CHIP_DEVICE_CONFIG_ESP32_BLE_ERROR_MIN + (CHIP_ERROR) bleErr;
```
Comment (yufengwangca) [inline src/platform/ESP32/nimble/BLEManagerImpl.cpp:459]: What is the return value of ErrorStr(err) if err is CHIP_DEVICE_CONFIG_ESP32_BLE_ERROR_MIN + (CHIP_ERROR) bleErr?

**Review by yufengwangca — CHANGES_REQUESTED** [review]

**Review by LuDuda — APPROVED** [review]
LGTM 👍 I would however consider removing `ret` temporary variable and use MapBLEError directly as suggested.

File: src/platform/ESP32/bluedroid/BLEManagerImpl.cpp (hunk 3 above)
Comment (sweetymhaiske) [reply src/platform/ESP32/bluedroid/BLEManagerImpl.cpp]: Done

File: src/platform/ESP32/bluedroid/BLEManagerImpl.cpp (hunk 4 above)
Comment (sweetymhaiske) [reply src/platform/ESP32/bluedroid/BLEManagerImpl.cpp]: Done

File: src/platform/ESP32/nimble/BLEManagerImpl.cpp (hunk 5 above)
Comment (sweetymhaiske) [reply src/platform/ESP32/nimble/BLEManagerImpl.cpp:455]: I have tried it with NimBLE error `BLE_HS_EUNKNOWN`.
```
CHIP_ERROR err = MapBLEError(BLE_HS_EUNKNOWN);

ChipLogError(DeviceLayer, "Return Value for ErrorStr(BLE_HS_EUNKNOWN) :- \n%s\n %d\n", ErrorStr(err), ErrorStr(err));
 ```
 **Output**
 ```
Return Value for ErrorStr(BLE_HS_EUNKNOWN) :-
Device Layer Error 11000017 (0x00A7D8D1)
1073512808
 ```

**Comment by sweetymhaiske** [comment]
> LGTM  I would however consider removing `ret` temporary variable and use MapBLEError directly as suggested.

Removed it.

File: src/platform/ESP32/nimble/BLEManagerImpl.cpp [hunk 6]
```diff
@@ -426,9 +426,37 @@ bool BLEManagerImpl::SendReadResponse(BLE_CONNECTION_OBJECT conId, BLE_READ_REQU
 
 void BLEManagerImpl::NotifyChipConnectionClosed(BLE_CONNECTION_OBJECT conId) {}
 
+CHIP_ERROR BLEManagerImpl::MapBLEError(int bleErr)
```
Comment (dhrishi) [inline src/platform/ESP32/nimble/BLEManagerImpl.cpp:429]: There are 2 types of APIs for NimBLE: Espressif added nimble APIs that return esp_err_t and upstream nimble APIs that return int. You seem to have combined their handling in a single API. See below `BLE_HS_EMSGSIZE` is from nimble and   `ESP_OK` and `ESP_ERR_INVALID_ARG` are from esp-nimble. For now, this should be fine

**Review by dhrishi — APPROVED** [review]

**Review by jmartinez-silabs — APPROVED** [review]

**Review by yufengwangca — APPROVED** [review]
